src/Symfony/Component/EventDispatcher/Attribute/AsEventListener.php

```diff
@@ -24,6 +24,8 @@ public function __construct(
         public ?string $method = null,
         public int $priority = 0,
         public ?string $dispatcher = null,
+        public ?string $before = null,
```

Comment:
indeed, I was wondering about introducing a value object
but it is actually incompatible with @kbond's [comment](https://github.com/symfony/symfony/pull/50687#pullrequestreview-1483529449) about using "quasi callable" which would be a nice DX addition (PHPStorm would rename the class/method in case of refactoring). I wish first class callable would be allowed in attributes definition :disappointed: 

Since `before` and `after` are incompatible, I'm wondering if I should not merge them, ie:
```php
#[AsEventListener(
    event: KernelEvents::REQUEST,
    beforeAfter: BeforeAfter::before('router_listener::onKernelRequest()')
)]
```
but I cannot find a good naming for this. maybe `PriorityHook`?

Reply:
> Could it at some point be an array shaped like: `array{service: string, method=: string}` (or even a DTO) to get rid of the string convention?

It should be `class-string|array{class?: class-string, service?: string, method?: string}`

---

This `priorityHook` is more verbose. There is an exception at compile-time if both `before` and `after` are used, so the current proposition is fine.

Reply:
even if first class callable was available in attributes, this would not be usable here: to get a callable for the RoutingListener `onKernelRequest` method, you need the *instance* of the RoutingListener as it is not a static method. `RoutingListener::onKernelRequest(...)` is a PHP error saying that the method cannot be called statically.

Reply:
yes, indeed you're right1

Reply:
Only specifying the class name or service name should cover most of the use-cases. The method name can be guessed when a class have only 1 method listening to an event. https://github.com/symfony/symfony/pull/50687#discussion_r1239639330

Reply:
I'm closing this one, discussion goes further in https://github.com/symfony/symfony/pull/50687#discussion_r1239639330